### ISSUE TITLE
Material not getting correct coupled neighbor values

### ISSUE THREAD

**Issue by friedmud** [idaholab/moose#5046]
TLDR: Neighbor material properties are messed up!

This is a straight up real bug that could have been affecting applications.

Material inherits from Coupleable.

When using DGKernels (or anything else that invokes the "neighbor" system: like Face Constraints and more) a Material will get created for "this" side of the interface and the "neighbor" side of the interface.

The Material that is created to represent the Material properties on the neighbor element is still inheriting from Coupleable... but that Coupleable interface doesn't know that it's supposed to be providing values from the neighbor element.  This means that _any_ coupled values in a neighbor material are actually the values from "this" element... NOT the neighbor!

I have a fix coming for this....

**Comment by YaqiWang** [comment]
Wow how do you notice this bug? Side material coupling with variables seems a really corner case.

**Comment by friedmud** [comment]
I found it while working on my element-centered finite difference code neutron diffusion code: https://github.com/friedmud/gardensnake

In the original implementation I was just telling the Material object where the edges of the baffle, reflector, fuel, etc. were.  That implementation was getting the exact same answers as others in my class (including the same number of power iterations).

Yesterday I was trying to switch to using an Auxiliary variable to define the "zones" (fuel, baffle, etc.).  When I did that my eigenvalues changed ever so slightly... and on one problem it took 3x the number of power iterations.  After many hours of digging further and further down... I finally figured it out....

When right on the edge between, say, reflector and fuel regions the new way was picking up the same region for BOTH sides of the edge in the Material objects that were feeding DGKernels.

Unbelievably subtle... but at least I found it eventually!  It was driving me crazy!

BTW: This isn't that obscure.  Anyone doing DG with a non-trivial problem would be impacted by this.  If you had _any_ amount of nonlinearity in your material properties (for instance a temperature dependent thermal conductivity in a DG heat conduction solve) then you would get a slightly wrong answer...

**Comment by YaqiWang** [comment]
You should try eigen executioners in moose for you fv eigen problems. We did use aux to store material Id but we didn't use DG with it. 😀

**Comment by friedmud** [comment]
I would: but this was a homework assignment... I didn't think it would be
very fair ;-) (Plus I need to learn!)

**Comment by permcody** [comment]
Yaqi this could be good for you. Derek might realize all of the
difficulties you might encounter when running eigenvalue problems and start
helping us design better interfaces.

Cody

**Comment by andrsd** [comment]
The values of a nodal variable on any edge will be the same for this element and the neighbor element, because the variable is continuos over the edge, right? So, why do you need `nodalSlnNeighbor` at el.?

Should not the fix in `Coupleble` be:
```
if (nodal)
  return var->nodalSln()
else
{
  if (_coupleable_neighbor)
    return var->slnNeighbor();
  else
    return var->sln();
}
```

**Comment by friedmud** [comment]
Remember: "neighbor" elements are not just used in DG.  The mechanism is also used in face-face contact (you also use it in Mortar, right?) and a few other things.  In some of those cases the nodes aren't shared so you would have different values...

**Comment by andrsd** [comment]
The mortar system works slightly differently. We have master values, slave values and mortar values. The mortar values are volumetric on lower dimensional mesh, master and slave values are obtained by reiniting either master or slave element at the projected quad-point on either side respectively, i.e. we compute volumetric values using special quadrature rule. So you do not really see values coming from `MooseVariable::neighbor_values`. Right now, one would get values from the mortar interface if one called `coupledValue`. We know about this "limitation", because currently we do only simple stuff like `u_master = u_slave`. The  'coupledValueNeighbor' is not available at all. In future, we would eventually need something like `coupledValue{Master|Slave}`. I do not currently know how to handle the case where we would bring an aux variable inside material, computed material property and then used it in the constraint. That would actually need 3 different types of storages, which we do not have.

As for the NodeFace CEDs, I know nodes are on the slave side and face values are on the master side, so the material property values are fine there and the slave side is not computing material properties at nodes, right?

You might be right, that NodeNode CEDs would need this, I did not look yet...

Is there anything else you can think of?
